### PR TITLE
feat: recurse into submodules

### PR DESCRIPTION
**Pull request by lehmanju**
### Usage

If `recurse_submodules` is set to `true`, you can include submodule commits into your changelog.

For example:

```toml
[git]
recurse_submodules = true

[changelog]
body = """
{% for submodule_path, commits in submodule_commits %}
    ### {{ submodule_path | upper_first }}
    {% for group, commits in commits | group_by(attribute="group") %}
        #### {{ group | upper_first }}
        {% for commit in commits %}
            - {{ commit.message | upper_first }}\
        {% endfor %}
    {% endfor %}
{% endfor %}\n
"""
```

### Implementation details

Adds a feature which allows git-cliff to recurse into submodules of a repository. Between two subsequent releseases in the main repository, all submodule commits in between are expanded. For each submodule, there is an entry in `submodule_commits` mapping the submodules path to a list of commits. This feature is off by default (recurse_submodules=false). Nested submodules are not supported.

### CODE REVIEW

**Comment by yudjinn** [comment]
Nice! So for a given commit range of the containing repo, this returns all submodules that changed within those commits, and their deltas between them?

**Comment by lehmanju** [comment]
Well, this iteration sort of actually works! However, the submodule path is not correct ...
[context.json](https://github.com/user-attachments/files/19132065/context.json)

I used openstack/openstack for testing btw, that repository does seem like the ultimate stresstest.

**Comment by lehmanju** [comment]
So that is fixed as well, only config option, documentation and code cleanup/review are missing.

**Comment by orhun** [comment]
This is some good stuff. Let me know when this is ready for review :)

File: git-cliff-core/Cargo.toml [hunk 1]
```diff
@@ -55,7 +55,7 @@ serde_json = "1.0.137"
 bincode = "2.0.0-rc.3"
 serde_regex = "1.1.0"
 tera = "1.20.0"
-indexmap = { version = "2.7.1", optional = true }
+indexmap = { version = "2.7.1", features = ["serde"] }
```
Comment (lehmanju) [inline git-cliff-core/Cargo.toml:58]: this isn't optional anymore since `Release` uses it for submodule commits. maybe use hashmap instead? my reasoning was that its used in `Repository` as well, so keep it consistent

File: git-cliff-core/src/changelog.rs [hunk 2]
```diff
@@ -835,6 +836,7 @@ mod test {
 				sort_commits:             Some(String::from("oldest")),
 				link_parsers:             None,
 				limit_commits:            None,
+				recurse_submodules:       None,
```
Comment (lehmanju) [inline git-cliff-core/src/changelog.rs:839]: not sure whether the name fits, it recurses only with depth 1 but I don't think there is a use case for higher depths.

File: git-cliff-core/src/commit.rs [hunk 3]
```diff
@@ -177,8 +177,8 @@ impl From<String> for Commit<'_> {
 }
 
 #[cfg(feature = "repo")]
-impl<'a> From<&GitCommit<'a>> for Commit<'a> {
-	fn from(commit: &GitCommit<'a>) -> Self {
+impl From<&GitCommit<'_>> for Commit<'_> {
```
Comment (lehmanju) [inline git-cliff-core/src/commit.rs:180]: the lifetimes of `GitCommit` and `Commit` are not related

File: git-cliff-core/src/release.rs [hunk 4]
```diff
@@ -31,35 +32,37 @@ use serde_json::value::Value;
...
-	pub timestamp:  i64,
+	pub timestamp:         i64,
 	/// Previous release.
-	pub previous:   Option<Box<Release<'a>>>,
+	pub previous:          Option<Box<Release<'a>>>,
 	/// Repository path.
-	pub repository: Option<String>,
+	pub repository:        Option<String>,
+	/// Submodule commits.
+	pub submodule_commits: Option<IndexMap<String, Vec<Commit<'a>>>>,
```
Comment (lehmanju) [inline git-cliff-core/src/release.rs:51]: Map of relative submodule path to a list of commits.

**Review by lehmanju — COMMENTED** [review]
not sure about some naming conventions and what should be tested/where.

File: git-cliff-core/src/release.rs (hunk 4 above)
Comment (orhun) [reply git-cliff-core/src/release.rs:51]: Can you add a comment explaining what the `IndexMap` contains? It's not clear to me what that `String` is supposed to be.

File: git-cliff-core/src/repo.rs [hunk 5]
```diff
@@ -90,6 +93,14 @@ impl Repository {
 		Ok(path)
 	}
 
+	/// Returns the initial path of the repository.
+	///
+	/// In case of a submodule this is the relative path to the toplevel
+	/// repository.
+	pub fn initial_path(&self) -> &PathBuf {
```
[the remark this answers is not in the document]
Comment (orhun) [reply git-cliff-core/src/repo.rs:100]: I recommend the following:

- Rename the existing `path` function to `root_path` since I think that's more fitting.
- Use `path` instead of `initial_path` here.

WDYT?

File: git-cliff/Cargo.toml [hunk 6]
```diff
@@ -60,7 +66,7 @@ url.workspace = true
 pathdiff = "0.2.3"
 
 [dependencies.git-cliff-core]
-version = "2.8.0" # managed by release.sh
+version = "2.8.0"          # managed by release.sh
```
Comment (orhun) [inline git-cliff/Cargo.toml:69]: Revert these formatting changes pls

File: git-cliff/src/lib.rs [hunk 7]
```diff
@@ -77,6 +78,46 @@ fn check_new_version() {
...
+	let first_commit = release
+		.commits
+		.first()
+		.map(|commit| &commit.id)
+		.and_then(|commit_id| repository.find_commit(commit_id));
+	let last_commit = release
+		.commits
+		.last()
+		.map(|commit| &commit.id)
+		.and_then(|commit_id| repository.find_commit(commit_id));
```
[the remark this answers is not in the document]
Comment (orhun) [reply git-cliff/src/lib.rs:95]: Yup, the ordering should be like that and it does not change. Not sure what you mean by the second part, does that affect us?

File: git-cliff-core/Cargo.toml (hunk 1 above)
Comment (orhun) [reply git-cliff-core/Cargo.toml:58]: If we use `hashmap` then the items are not ordered which is probably not something that we want.

Also why enable the `serde` feature?

File: git-cliff-core/src/changelog.rs (hunk 2 above)
Comment (orhun) [reply git-cliff-core/src/changelog.rs:839]: I think that's fine (also sounds cool :D)

**Review by orhun — COMMENTED** [review]
Looks nice!

Some things to do:

- Add [fixture tests](https://github.com/orhun/git-cliff/tree/main/.github/fixtures) to ensure the functionality
- Update documentation about the new configuration option

**Comment by lehmanju** [comment]
Thanks for the review! I'll have time to address your comments in about a week.

File: git-cliff-core/Cargo.toml (hunk 1 above)
Comment (lehmanju) [reply git-cliff-core/Cargo.toml:58]: serde is needed now because `Release`  needs to be serializable and as a consequence `indexmap` .

File: git-cliff/src/lib.rs [hunk 8]
```diff
@@ -77,6 +81,56 @@ fn check_new_version() {
...
+	let first_commit = release
+		.previous
+		.as_ref()
+		.and_then(|previous_release| previous_release.commit_id.clone())
+		.and_then(|commit_id| repository.find_commit(&commit_id));
+	let last_commit = release
+		.commit_id
+		.clone()
+		.and_then(|commit_id| repository.find_commit(&commit_id));
+	let commit_range = first_commit.zip(last_commit);
```
Comment (lehmanju) [inline git-cliff/src/lib.rs:99]: This commit range generation was initially wrong, I needed to use the tag commit id from the previous and current release.

File: git-cliff/src/lib.rs [hunk 9]
```diff
@@ -300,6 +355,9 @@ fn process_repository<'a>(
 			previous_release.previous = None;
 			release.previous = Some(Box::new(previous_release));
 			previous_release = release.clone();
+			if recurse_submodules {
```
Comment (lehmanju) [inline git-cliff/src/lib.rs:358]: Because previous release is needed, I moved this to the bottom of the for loop. Else it wouldn't be initialized.

File: git-cliff-core/src/release.rs [hunk 10]
```diff
@@ -31,35 +33,40 @@ use serde_json::value::Value;
...
-	pub previous:   Option<Box<Release<'a>>>,
+	pub previous:          Option<Box<Release<'a>>>,
 	/// Repository path.
-	pub repository: Option<String>,
+	pub repository:        Option<String>,
+	/// Submodule commits.
+	///
+	/// Maps submodule path to a list of commits.
+	#[serde(rename = "submodule_commits")]
+	pub submodule_commits: Option<HashMap<String, Vec<Commit<'a>>>>,
```
Comment (lehmanju) [inline git-cliff-core/src/release.rs:55]: Changed to Hashmap because indexmap serde implementation doesn't preserve order anyway and order doesn't matter.

File: git-cliff-core/src/changelog.rs [hunk 11]
```diff
@@ -186,6 +145,58 @@ impl<'a> Changelog<'a> {
 		Ok(())
 	}
 
+	fn process_commit_list(
```
Comment (lehmanju) [inline git-cliff-core/src/changelog.rs:148]: This is new, because commit preprocessing now needs to be done for release commits AND submodule commits.

**Review by lehmanju — COMMENTED** [review]
Did a bunch of fixes:
- added fixture test
- fixed first/last commit bug (submodule commit range now correct)
- fixed commit preprocessing for submodule commits

@orhun would appreciate another review.

**Comment by lehmanju** [comment]
There does seem to be something wrong with the fixture tests in general: https://github.com/orhun/git-cliff/actions/runs/13907737767/job/38914649566#step:3:913

File: git-cliff-core/src/release.rs (hunk 10 above)
Comment (lehmanju) [reply git-cliff-core/src/release.rs]: removed the Option<> wrapping and used an empty hashmap instead. results in a cleaner API during templating.

**Comment by lehmanju** [comment]
The more I test this in a real scenario the more corner cases I find:
unreleased and first release don't contain any submodule commits although they should

**Review by orhun — COMMENTED** [review]
Looking good! Left some comments that I believe will make the implementation/documentation better.

The fixtures should be fixed in #1099 - ignore the ones that are failing for now.

File: website/docs/templating/context.md [hunk 12]
```diff
@@ -64,6 +64,15 @@ following context is generated to use for templating:
   "commit_id": "a440c6eb26404be4877b7e3ad592bfaa5d4eb210 (release commit)",
   "timestamp": 1625169301,
   "repository": "/path/to/repository",
+  "submodule_commits": {
+    "<relative/submodule_path/in/repository>": [
```
[the remark this answers is not in the document]
Comment (lehmanju) [reply website/docs/templating/context.md:68]: Do you mean, if git-cliff is invoked from a repository subdirectory that is not a submodule or from the submodule itself?

[the remark this answers is not in the document]
Comment (lehmanju) [reply website/docs/templating/context.md:68]: added another fixture test for include-path. if git-cliff is invoked inside a submodule, it will only be able to process the submodule itself. if git-cliff is invoked inside a normal monorepo directory, it should work as expected. the path that is used to find the submodule is gathered from the commit itself, so it doesn't depend on where git-cliff is invoked.

**Comment by lehmanju** [comment]
@orhun If CI turns out good (which it should), this is ready for a (hopefully) final review.

- docs updated
- fixture tests improved
- edge cases (submodule added/removed) working
- releases without commit id working

File: .github/fixtures/test-submodules-include-path/cliff.toml
Comment (orhun) [inline .github/fixtures/test-submodules-include-path/cliff.toml:1]: This fixture config can now be further simplified thanks to #1087 

Can you remove the default values and just leave the ones that affect the fixture result?

File: .github/fixtures/test-submodules/cliff.toml
Comment (orhun) [inline .github/fixtures/test-submodules/cliff.toml:1]: This config can be simplified, see my comment above.

**Review by orhun — COMMENTED** [review]
Thank you, this looks great! Only a couple of tweaks left:

- I left a review about the fixture configs, they can be simplified.
- Can you update the default config (`config/cliff.toml`) and include the `recurse_submodules` there?
- Lastly, we need to rebase on `main` :)

And sorry for late review, I've been traveling...

**Comment by lehmanju** [comment]
Any chance to get a new release by Monday? This would really help me at my job.

**Comment by orhun** [comment]
I was already planning a release soon, I'll try to push it sooner :)
